keylime/src/device_id.rs

```diff
@@ -0,0 +1,535 @@
+        let detected = &crypto::match_cert_to_template(iak_cert)
+            .map_err(DeviceIDBuilderError::TemplateFromCert)?;
+
+        let template = self.iak_template.unwrap_or("").trim();
+        let asym_alg = self.iak_asym_alg.unwrap_or("").trim();
+        let hash_alg = self.iak_hash_alg.unwrap_or("").trim();
+
+        tpm::get_idevid_template(detected, template, asym_alg, hash_alg)
+            .map_err(DeviceIDBuilderError::Template)
+    }
+
+    /// Get the IDevID template
+    ///
+    /// If configured to detect, get the template from the IDevID certificate
+    /// Otherwise, construct the template from the provided algorithms
+    fn get_idevid_template(
+        &mut self,
+    ) -> Result<(AsymmetricAlgorithm, HashingAlgorithm), DeviceIDBuilderError>
+    {
+        let idevid_cert = self.get_idevid_cert()?;
+        let detected = &crypto::match_cert_to_template(idevid_cert)
+            .map_err(DeviceIDBuilderError::TemplateFromCert)?;
+
+        let template = self.idevid_template.unwrap_or("").trim();
+        let asym_alg = self.idevid_asym_alg.unwrap_or("").trim();
+        let hash_alg = self.idevid_hash_alg.unwrap_or("").trim();
+
+        tpm::get_idevid_template(detected, template, asym_alg, hash_alg)
+            .map_err(DeviceIDBuilderError::Template)
+    }
+
+    /// Get the IAK from the given handle and password or recreate following the given template
+    ///
+    /// If a handle has been set, try to obtain the IAK from the handle
+    /// If there is a configured IAK password, add the password to the handle
+    ///
+    /// If the handle is empty, recreate the IAK using the provided algorithms
+    fn get_iak(
+        &mut self,
+        tpm_ctx: &mut tpm::Context,
+    ) -> Result<IAKResult, DeviceIDBuilderError> {
+        let (asym_alg, hash_alg) = self.get_iak_template()?;
+        match self.iak_handle {
+            Some(handle) => {
+                if handle.trim().is_empty() {
+                    info!("Recreating IAK.");
+                    tpm_ctx
+                        .create_iak(asym_alg, hash_alg)
+                        .map_err(DeviceIDBuilderError::IAKCreate)
+                } else {
+                    let password = self.iak_password.unwrap_or("").trim();
+                    info!("Collecting persisted IAK.");
+                    tpm_ctx
+                        .iak_from_handle(handle.trim(), password)
+                        .map_err(DeviceIDBuilderError::IAKFromHandle)
+                }
+            }
+            None => Err(DeviceIDBuilderError::IAKHandleNotSet),
+        }
+    }
+
+    /// Get the IDevID from the given handle and password or recreate following the given template
+    ///
+    /// If a handle has been set, try to obtain the IDevID from the handle
+    /// If there is a configured IDevID password, add the password to the handle
+    ///
+    /// If the handle is empty, recreate the IDevId using the provided algorithms
+    fn get_idevid(
```

Comment:
I think if this is recreated here we need to flush it as otherwise we will have 4 keys at once (IAK, IDevID, AK, EK) which can cause an out of memory TPM error.

Reply:
Since the IDevID key is not needed for the AK certification later, do you think we should flush it immediately after checking that the certificate matches its public key and drop it from the `DeviceID` structure?

Maybe the `DeviceID` structure should contain the IAK handle, IAK pubkey, IAK certificate, IDevID pubkey, and IDevID certificate instead of holding the `IAKResult` and `IDevIDResult`.

Something like:
```
#[derive(Debug)]
pub struct DeviceID {
    pub iak_handle: KeyHandle,
    pub iak_pub: TssPublic,
    pub iak_cert: X509,
    pub idevid_pub: TssPublic,
    pub idevid_cert: X509,
}
```
We could also flush the IAK handle after certifying AK with it as it is not necessary later.

Reply:
We don't want to attempt to flush if the handle is from the config rather than from us regenerating the key. So I think we can flush the IDevID immediately after regenerating as long as we have collected the pub value if we are going to check against the cert.

Same for the IAK (we only want to flush after the certify if we regenerated it). I suppose you could check to see if there is a value in the handle value in config before deciding to flush?

That structure looks like all we need at the moment (but I think we maybe want the certs to be options, see other comments).

Reply:
Not the most elegant solution, but I added a `is_persistent` flag in the `IAKResult` and `IDevIDResult` to indicate if the handles should be flushed after use. Reading the configuration was not an option since the goal is to move the code to the `keylime` library, detached from the `keylime-agent` application.